llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp

```diff
@@ -5360,10 +5368,20 @@ bool SelectionDAG::isKnownNeverZero(SDValue Op, unsigned Depth) const {
     return isKnownNeverZero(Op.getOperand(1), Depth + 1) ||
            isKnownNeverZero(Op.getOperand(0), Depth + 1);
 
-    // TODO for smin/smax: If either operand is known negative/positive
+    // For smin/smax: If either operand is known negative/positive
     // respectively we don't need the other to be known at all.
   case ISD::SMAX:
+    if (computeKnownBits(Op.getOperand(1), Depth + 1).isStrictlyPositive() ||
+        computeKnownBits(Op.getOperand(0), Depth + 1).isStrictlyPositive())
+      return true;
+    return isKnownNeverZero(Op.getOperand(1), Depth + 1) &&
+           isKnownNeverZero(Op.getOperand(0), Depth + 1);
   case ISD::SMIN:
+    if (computeKnownBits(Op.getOperand(1), Depth + 1).isNegative() ||
+        computeKnownBits(Op.getOperand(0), Depth + 1).isNegative())
```

Comment:
Both here and above, think you should cache the result of `computeKnownBits` so you can check if the `KnownBits` is trivially non-zero and potentially avoid 1/2 recursive calls through `isKnownNeverZero`.

Reply:
Not sure how I would do that.

Reply:
Maybe pull out the computeKnownBits calls so we can also test:
```cpp
if (computeKnownBits(Op.getOperand(1), Depth + 1).isNonZero() &&
    computeKnownBits(Op.getOperand(0), Depth + 1).isNonZero())
```